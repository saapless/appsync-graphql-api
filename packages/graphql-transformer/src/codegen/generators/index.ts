export { DynamoDbGenerator } from "./DynamoDBGenerator";
export { TypeGenerator } from "./type";
