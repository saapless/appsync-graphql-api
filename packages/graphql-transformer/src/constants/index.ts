export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
