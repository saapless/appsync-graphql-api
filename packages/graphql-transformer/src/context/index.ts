export * from "./TransformerContext";
export type {
  DescriptorOperation,
  Key,
  KeyOperator,
  KeyValue,
  OperationActionType,
  ResolverDescriptor,
  ReturnType,
} from "./ResolverLoader";
