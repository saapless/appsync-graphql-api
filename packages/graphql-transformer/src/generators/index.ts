export { ResolverGeneratorBase } from "./ResolverGeneratorBase";
export { DynamoDbGenerator } from "./DynamoDBGenerator";
export { NoneGenerator } from "./NoneGenerator";
export { SchemaTypesGenerator } from "./SchemaTypesGenerator";
export { ResolverTypesGenerator } from "./ResolverTypesGenerator";
