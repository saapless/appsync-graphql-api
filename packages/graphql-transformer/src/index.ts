export * from "./transformer";
export { definitionFromSource } from "./utils";
