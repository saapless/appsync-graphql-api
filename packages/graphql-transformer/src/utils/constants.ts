/**
 * DIRECTIVES
 */
export const AUTH_DIRECTIVE = "auth";

export const AWS_AUTH_DIRECTIVE = "aws_auth";
export const AWS_API_KEY_DIRECTIVE = "aws_api_key";
export const AWS_IAM_DIRECTIVE = "aws_iam";
export const AWS_OIDC_DIRECTIVE = "aws_oidc";
export const AWS_LAMBDA_DIRECTIVE = "aws_lambda";
export const AWS_COGNITO_USER_POOLS_DIRECTIVE = "aws_cognito_user_pools";
export const AWS_SUBSCRIBE_DIRECTIVE = "aws_subscribe";

export const HAS_ONE_DIRECTIVE = "hasOne";
export const HAS_MANY_DIRECTIVE = "hasMany";

export const DATA_SOURCE_DIRECTIVE = "dataSource";
export const RESOLVER_DIRECTIVE = "resolver";

export const MODEL_DIRECTIVE = "model";

export const SERVER_ONLY_DIRECTIVE = "serverOnly";
export const CLIENT_ONLY_DIRECTIVE = "clientOnly";
export const READ_ONLY_DIRECTIVE = "readOnly";
export const WRITE_ONLY_DIRECTIVE = "writeOnly";
export const FILTER_ONLY_DIRECTIVE = "filterOnly";

export const DEFAULT_ID_FIELD = "id";
export const DEFAULT_SK_FIELD = "_sk";

export const DEFAULT_SOURCE_INDEX = "bySourceId";
export const DEFAULT_SOURCE_INDEX_FIELD = "sourceId";
export const DEFAULT_TARGET_INDEX = "byTargetId";
export const DEFAULT_TARGET_INDEX_FIELD = "targetId";
export const DEFAULT_TYPENAME_INDEX = "byTypename";
export const DEFAULT_TYPENAME_INDEX_FIELD = "__typename";
