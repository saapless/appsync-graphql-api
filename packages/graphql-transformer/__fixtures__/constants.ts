import { DataSourceManagerConfig } from "../src/context/DataSourceManager";

export const TEST_DS_CONFIG = {
  primaryDataSourceName: "TestDataSource",
  dataSources: {
    TestDataSource: {
      type: "DYNAMO_DB",
      config: {
        tableName: "TestTable",
      },
    },
    NoneDataSource: {
      type: "NONE",
    },
  },
} satisfies DataSourceManagerConfig;
