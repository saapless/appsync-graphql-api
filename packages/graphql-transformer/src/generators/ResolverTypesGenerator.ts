import ts from "typescript";
import { TransformerContext } from "../context";
import { FieldLoaderDescriptor, pascalCase } from "../utils";
import { ObjectNode } from "../definition";
import { GeneratorBase } from "./GeneratorBase";

export class ResolverTypesGenerator extends GeneratorBase {
  constructor(context: TransformerContext) {
    super(context);
  }

  private _ddbDefaults() {
    this._definitions.push(
      // Query
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("DynamoDBQueryResult"),
        [
          ts.factory.createTypeParameterDeclaration(
            undefined,
            "T",
            ts.factory.createTypeReferenceNode("Key"),
            ts.factory.createTypeReferenceNode("Key")
          ),
        ],
        ts.factory.createTypeLiteralNode([
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("items"),
            undefined,
            ts.factory.createArrayTypeNode(ts.factory.createTypeReferenceNode("T"))
          ),
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("nextToken"),
            undefined,
            ts.factory.createUnionTypeNode([
              ts.factory.createTypeReferenceNode("string"),
              ts.factory.createTypeReferenceNode("null"),
            ])
          ),
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("scannedCount"),
            undefined,
            ts.factory.createTypeReferenceNode("number")
          ),
        ])
      ),
      // TrasactionCancellationReason
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("TrasactionCancellationReason"),
        undefined,
        ts.factory.createTypeLiteralNode([
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("type"),
            undefined,
            ts.factory.createTypeReferenceNode("string")
          ),
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("message"),
            undefined,
            ts.factory.createTypeReferenceNode("string")
          ),
        ])
      ),
      // Batch
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("DynamoDBBatchResult"),
        [
          ts.factory.createTypeParameterDeclaration(
            undefined,
            "T",
            ts.factory.createTypeReferenceNode("Key"),
            ts.factory.createTypeReferenceNode("Key")
          ),
        ],
        ts.factory.createTypeLiteralNode([
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("data"),
            undefined,
            ts.factory.createTypeLiteralNode([
              ts.factory.createIndexSignature(
                undefined,
                [
                  ts.factory.createParameterDeclaration(
                    undefined,
                    undefined,
                    ts.factory.createIdentifier("K"),
                    undefined,
                    ts.factory.createTypeReferenceNode("string")
                  ),
                ],
                ts.factory.createArrayTypeNode(
                  ts.factory.createUnionTypeNode([
                    ts.factory.createTypeReferenceNode("T"),
                    ts.factory.createTypeReferenceNode("null"),
                  ])
                )
              ),
            ])
          ),
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("unprocessedKeys"),
            undefined,
            ts.factory.createTypeLiteralNode([
              ts.factory.createIndexSignature(
                undefined,
                [
                  ts.factory.createParameterDeclaration(
                    undefined,
                    undefined,
                    ts.factory.createIdentifier("K"),
                    undefined,
                    ts.factory.createTypeReferenceNode("string")
                  ),
                ],
                ts.factory.createArrayTypeNode(ts.factory.createTypeReferenceNode("Key"))
              ),
            ])
          ),
        ])
      ),
      // TransactGet
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("DynamoDBTransactGetResult"),
        [
          ts.factory.createTypeParameterDeclaration(
            undefined,
            "T",
            ts.factory.createTypeReferenceNode("Key"),
            ts.factory.createTypeReferenceNode("Key")
          ),
        ],
        ts.factory.createUnionTypeNode([
          ts.factory.createTypeLiteralNode([
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("items"),
              undefined,
              ts.factory.createArrayTypeNode(ts.factory.createTypeReferenceNode("T"))
            ),
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("cancellationReasons"),
              undefined,
              ts.factory.createTypeReferenceNode("null")
            ),
          ]),
          ts.factory.createTypeLiteralNode([
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("items"),
              undefined,
              ts.factory.createTypeReferenceNode("null")
            ),
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("cancellationReasons"),
              undefined,
              ts.factory.createArrayTypeNode(
                ts.factory.createTypeReferenceNode("TrasactionCancellationReason")
              )
            ),
          ]),
        ])
      ),
      // TransactWrite
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("DynamoDBTransactWriteResult"),
        undefined,
        ts.factory.createUnionTypeNode([
          ts.factory.createTypeLiteralNode([
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("keys"),
              undefined,
              ts.factory.createArrayTypeNode(ts.factory.createTypeReferenceNode("Key"))
            ),
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("cancellationReasons"),
              undefined,
              ts.factory.createTypeReferenceNode("null")
            ),
          ]),
          ts.factory.createTypeLiteralNode([
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("keys"),
              undefined,
              ts.factory.createTypeReferenceNode("null")
            ),
            ts.factory.createPropertySignature(
              undefined,
              ts.factory.createIdentifier("cancellationReasons"),
              undefined,
              ts.factory.createArrayTypeNode(
                ts.factory.createTypeReferenceNode("TrasactionCancellationReason")
              )
            ),
          ]),
        ])
      )
    );
  }

  private _defaults() {
    this._definitions.push(
      ts.factory.createImportDeclaration(
        undefined,
        ts.factory.createImportClause(
          false,
          undefined,
          ts.factory.createNamedImports([
            ts.factory.createImportSpecifier(
              true,
              undefined,
              ts.factory.createIdentifier("Context")
            ),
          ])
        ),
        ts.factory.createStringLiteral("@aws-appsync/utils")
      ),
      ts.factory.createImportDeclaration(
        undefined,
        ts.factory.createImportClause(
          true,
          undefined,
          ts.factory.createNamespaceImport(ts.factory.createIdentifier("Schema"))
        ),
        ts.factory.createStringLiteral("./schema-types")
      ),
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier("Key"),
        undefined,
        ts.factory.createTypeLiteralNode([
          ts.factory.createPropertySignature(
            undefined,
            ts.factory.createIdentifier("id"),
            undefined,
            ts.factory.createTypeReferenceNode(ts.factory.createIdentifier("string"))
          ),
        ])
      )
    );

    this._ddbDefaults();
  }

  private _args(loader: FieldLoaderDescriptor) {
    const node = this._context.document.getNode(loader.typeName);

    if (node instanceof ObjectNode && Boolean(node.getField(loader.fieldName)?.arguments?.length)) {
      return ts.factory.createTypeReferenceNode(
        `Schema.${pascalCase(loader.typeName, loader.fieldName, "args")}`
      );
    }

    return ts.factory.createTypeReferenceNode("Record", [
      ts.factory.createTypeReferenceNode("string"),
      ts.factory.createTypeReferenceNode("unknown"),
    ]);
  }

  private _source(loader: FieldLoaderDescriptor) {
    if (loader.typeName === "Query" || loader.typeName === "Mutation") {
      return ts.factory.createTypeReferenceNode("undefined");
    }

    return ts.factory.createTypeReferenceNode(`Schema.${loader.typeName}`);
  }

  private _result(loader: FieldLoaderDescriptor) {
    switch (loader.action.type) {
      case "getItem":
      case "putItem":
      case "updateItem":
      case "upsertItem":
      case "removeItem":
        return ts.factory.createTypeReferenceNode(`Schema.${loader.targetName}`);
      case "queryItems":
        return ts.factory.createTypeReferenceNode("DynamoDBQueryResult", [
          ts.factory.createTypeReferenceNode(`Schema.${loader.targetName}`),
        ]);
      case "batchGetItems":
        return ts.factory.createTypeReferenceNode("DynamoDBBatchResult", [
          ts.factory.createTypeReferenceNode(`Schema.${loader.targetName}`),
        ]);
      default:
        return ts.factory.createTypeReferenceNode("unknown");
    }
  }

  private _field(loader: FieldLoaderDescriptor) {
    const args = this._args(loader);
    const source = this._source(loader);
    const result = this._result(loader);

    this._definitions.push(
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier(
          pascalCase(loader.typeName, loader.fieldName, "req", "context")
        ),
        undefined,
        ts.factory.createTypeReferenceNode(ts.factory.createIdentifier("Context"), [
          // Args
          args,
          // Stash
          ts.factory.createTypeReferenceNode("Record", [
            ts.factory.createTypeReferenceNode("string"),
            ts.factory.createTypeReferenceNode("unknown"),
          ]),
          // Prev
          ts.factory.createTypeReferenceNode("undefined"),
          // Source
          source,
          // Result
          ts.factory.createTypeReferenceNode("undefined"),
        ])
      ),
      ts.factory.createTypeAliasDeclaration(
        [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
        ts.factory.createIdentifier(
          pascalCase(loader.typeName, loader.fieldName, "res", "context")
        ),
        undefined,
        ts.factory.createTypeReferenceNode(ts.factory.createIdentifier("Context"), [
          // Args
          args,
          // Stash
          ts.factory.createTypeReferenceNode("Record", [
            ts.factory.createTypeReferenceNode("string"),
            ts.factory.createTypeReferenceNode("unknown"),
          ]),
          // Prev
          ts.factory.createTypeReferenceNode("undefined"),
          // Source
          source,
          // Result
          result,
        ])
      )
    );
  }

  // private _function(loader: PipelineFunctionLoaderDescriptor) {}

  public generate(filename: string): string {
    this._defaults();

    for (const fieldLoader of this._context.loader.getAllFieldLoaders()) {
      this._field(fieldLoader);
    }

    // TODO: enable if needed
    // for (const pipelineFunction of this._context.loader.getAllFunctionLoaders()) {
    //   this._function(pipelineFunction);
    // }

    return this._printDefinitions(filename);
  }
}
