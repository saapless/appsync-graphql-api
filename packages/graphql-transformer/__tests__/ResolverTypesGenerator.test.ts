import { TEST_DS_CONFIG } from "../__fixtures__/constants";
import { TransformerContext } from "../src/context";
import { DocumentNode } from "../src/definition";
import { ResolverTypesGenerator } from "../src/generators/ResolverTypesGenerator";

const context = new TransformerContext({
  document: DocumentNode.fromSource(/* GraphQL */ `
    type User {
      id: ID!
      name: String!
    }

    input CreateUserInput {
      id: ID
      name: String
    }

    input UpdateUserInput {
      id: ID!
      name: String
    }

    type Query {
      me: User
    }

    type Mutation {
      createUser(input: CreateUserInput!): User
      updateUser(input: UpdateUserInput!): User
      deleteUser(id: ID!): User
    }
  `),
  dataSourceConfig: TEST_DS_CONFIG,
});

describe("SchemaTypesGenerator", () => {
  beforeAll(() => {
    context.loader.setFieldLoader("Query", "me", {
      targetName: "User",
      action: {
        type: "getItem",
        key: {
          id: { ref: "identity.sub" },
        },
      },
      returnType: "result",
    });
    context.loader.setFieldLoader("Mutation", "createUser", {
      targetName: "User",
      action: {
        type: "putItem",
        key: { id: { ref: "identity.sub" } },
      },
      returnType: "result",
    });
    context.loader.setFieldLoader("Mutation", "updateUser", {
      targetName: "User",
      action: {
        type: "updateItem",
        key: { id: { ref: "identity.sub" } },
      },
      returnType: "result",
    });
    context.loader.setFieldLoader("Mutation", "deleteUser", {
      targetName: "User",
      action: {
        type: "removeItem",
        key: { id: { ref: "args.id" } },
      },
      returnType: "result",
    });
  });

  it("genrates schema types", () => {
    const generator = new ResolverTypesGenerator(context);

    const types = generator.generate("resolver-types.ts");
    expect(types).toMatchSnapshot();
  });
});
