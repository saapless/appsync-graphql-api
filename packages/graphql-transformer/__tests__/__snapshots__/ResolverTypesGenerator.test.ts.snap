// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolverTypesGenerator generates resolver context types 1`] = `
"import { type Context } from "@aws-appsync/utils";
import type * as Schema from "./schema-types";
export type Key = {
    id: string;
};
export type DynamoDBQueryResult<T extends Key = Key> = {
    items: T[];
    nextToken: string | null;
    scannedCount: number;
};
export type TrasactionCancellationReason = {
    type: string;
    message: string;
};
export type DynamoDBBatchResult<T extends Key = Key> = {
    data: {
        [K: string]: (T | null)[];
    };
    unprocessedKeys: {
        [K: string]: Key[];
    };
};
export type DynamoDBTransactGetResult<T extends Key = Key> = {
    items: T[];
    cancellationReasons: null;
} | {
    items: null;
    cancellationReasons: TrasactionCancellationReason[];
};
export type DynamoDBTransactWriteResult = {
    keys: Key[];
    cancellationReasons: null;
} | {
    keys: null;
    cancellationReasons: TrasactionCancellationReason[];
};
export type QueryMeReqContext = Context<Record<string, unknown>, Record<string, unknown>, undefined, undefined, undefined>;
export type QueryMeResContext = Context<Record<string, unknown>, Record<string, unknown>, undefined, undefined, Schema.User>;
export type MutationCreateUserReqContext = Context<Schema.MutationCreateUserArgs, Record<string, unknown>, undefined, undefined, undefined>;
export type MutationCreateUserResContext = Context<Schema.MutationCreateUserArgs, Record<string, unknown>, undefined, undefined, Schema.User>;
export type MutationUpdateUserReqContext = Context<Schema.MutationUpdateUserArgs, Record<string, unknown>, undefined, undefined, undefined>;
export type MutationUpdateUserResContext = Context<Schema.MutationUpdateUserArgs, Record<string, unknown>, undefined, undefined, Schema.User>;
export type MutationDeleteUserReqContext = Context<Schema.MutationDeleteUserArgs, Record<string, unknown>, undefined, undefined, undefined>;
export type MutationDeleteUserResContext = Context<Schema.MutationDeleteUserArgs, Record<string, unknown>, undefined, undefined, Schema.User>;
"
`;
